Elaborate Contract Basics based on new developer experience (#1101)
Agoric/documentation: main/guides/zoe/contract-basics.md

_We'll discuss [Far in more detail](../js-programming/far) later._

Putting it all together:

<<< @/../snippets/zoe/src/01-hello.js#contract

Let us save this code to a file named `01-hello.js` inside `src` directory. 
## Using, testing a contract

Agoric contracts are typically tested using the [ava](https://github.com/avajs/ava) framework. The test file begins with an `import @endo/init` to establish a [Hardened JavaScript](../js-programming/hardened-js) environment. We also import `E()` in order to make asynchronous method calls and `test` function from `ava`. _We'll talk more about [using `E()` for async method calls](../js-programming/eventual-send) later._ Following these `import` statements, we write a simple test that validates that the `greet` method works as expected.

Putting it all together:

<<< @/../snippets/zoe/contracts/test-zoe-01-hello.js#test-01-hello

Let's save this code in a file named `test-01-hello.js` in a `test` directory. Both `src` and `test` directories should lie in the same `contract` directory. Let us run the following command to execute the test:

```sh

> ava --match="contract greets by name"

Reviewer: oops! when I requested taking out `npx`, I should have offered a replacement:

```suggestion
yarn ava --match="contract greets by name"
```

I don't expect that invoking `ava` on its own will work unless somebody has done `npm install -g ava`.

not critical, I suppose

Author: Thanks @dckc. I am adding "follow the instructions there to install ava" right next to where we provide link for "ava". Probably we will fix this with something better soon.

# Smart Contract Basics

Before we look at how to make a contract such as the one in [the

> basic dapp](../getting-started/) in the previous section, let's cover some basics by writing a simple contract that returns a greetings message. We will simply call it _greetings smart contract_. 

Reviewer: It's supposed to be a hello world contract. Can we call it that, please?

```suggestion
basic dapp](../getting-started/) in the previous section, let's cover some basics by writing a simple hello-world contract. 
```

Author: Thank you for saying that. Fixed.

> A contract is defined by a JavaScript module that exports a `start` function. For our greetings smart contract, the declaration of `start` function looks like this:

Reviewer: ```suggestion
A contract is defined by a JavaScript module that exports a `start` function. For our hello-world smart contract, the declaration of `start` function looks like this:
```

and likewise throughout

Author: Fixed.

<<< @/../snippets/zoe/src/01-hello.js#start

> For the greetings smart contract, we will have a simple `greet` function apart from `start` function. The `greet` function takes a string as a parameter (for example, name of the person calling the function) and returns a customized greeting message.

Reviewer: Our audience is expected to be able to read JS. I'm not sure what the "The `greet` function ... returns a ... message" sentence tells them about our smart contract platform.

Author: It is one line summary of the contract - can we keep it?

Reviewer: ok

<<< @/../snippets/zoe/src/01-hello.js#greet

The `greet` function, along with any other public function, must be made accessible through the `publicFacet` of the contract. The `start` function returns an object with a `publicFacet` property. In the greeting contract, the `start` function exposes the `greet` function by defining it as a method of the contract's `publicFacet`, as shown below:

<<< @/../snippets/zoe/src/01-hello.js#publicFacet

> We wrap the `greet` function in a `Far(...)` call to safely expose it as a remote object, accessible from outside the contract.

Reviewer: TECHNICAL:

```suggestion
We wrap the value of the `publicFacet` property in a `Far(...)` call to safely expose it as a remote object, accessible from outside the contract.
```

In the text before this change, the "it" in "We mark it Far" referred to the public facet object, not the greet function.

Author: Fixed.

> We also give it a suggestive interface name `Hello` for debugging.

Reviewer: TECHNICAL:

This phrasing suggests that giving a suggestive name is separate from wrapping with Far.

```suggestion
This also gives it a suggestive interface name `Hello` for debugging.
```

Author: Fixed.

_We'll discuss [Far in more detail](../js-programming/far) later._

Putting it all together:

<<< @/../snippets/zoe/src/01-hello.js#contract

Let us save this code to a file named `01-hello.js` inside `src` directory. 
## Using, testing a contract

> Before we deploy or start our contract, it is usually a good idea to run a few tests to validate its correctness. Agoric contracts are typically tested using the [ava](https://github.com/avajs/ava) framework. The test file begins with an `import @endo/init` to establish a [Hardened JavaScript](../js-programming/hardened-js) environment. We also import `E()` in order to make asynchronous method calls and `test` function from `ava`. _We'll talk more about [using `E()` for async method calls](../js-programming/eventual-send) later._ Following these `import` statements, we write a simple test that validates that the `greet` method works as expected.

Reviewer: Again, deployment seems out of scope.

Author: Removed. Fixed.

# Smart Contract Basics

Before we look at how to make a contract such as the one in [the

> basic dapp](../getting-started/) in the previous section, let's cover some basics by writing a simple contract that returns a greetings message. We will simply call it _greetings smart contract_. 

Author: I would like to, politely, disagree :-) I think the requested change do make life easier for a new dev.

> basic dapp](../getting-started/) in the previous section, let's cover some basics by writing a simple contract that returns a greetings message. We will simply call it _greetings smart contract_. 

Reviewer: What I heard in office hours is that your position is not just based on your opinion, but on several developer's experience with learning the platform. I can go with that.

Ideally, next time, open an issue with some details about that experience; that issue can be prioritized and serve as justification for a PR such as this one.

> basic dapp](../getting-started/) in the previous section, let's cover some basics by writing a simple contract that returns a greetings message. We will simply call it _greetings smart contract_. 

Author: Fixed.

A contract is defined by a JavaScript module that exports a `start` function. For our greetings smart contract, the declaration of `start` function looks like this:

<<< @/../snippets/zoe/src/01-hello.js#start

> Apart from `start`, we can have extra functions in our contract. These functions may be defined outside or within the `start` function depending on whether they need access to the variables defined in the `start` function.  For the greetings smart contract, we will have a simple `greet` function (outside `start`) that takes a string as a parameter (for example, name of the person calling the function) and returns a customized greeting message.

Reviewer 2: > In addition to start, we can include additional functions in our contract. These functions can be defined either outside or within the start function, depending on whether they need access to variables defined in start. 

This can be reduced/concise for users who probably might already be familiar with basic JavaScript concepts.

Author: Let us talk about it. I am a bit inclined to err towards redundancy in the first tutorial.

Author: I made it short after our discussion, as below: 

For the greetings smart contract, we will have a simple `greet` function apart from `start` function. The `greet` function takes a string as a parameter (for example, name of the person calling the function) and returns a customized greeting message.

<<< @/../snippets/zoe/src/01-hello.js#greet

> The `greet` function, along with any other function besides `start`, must be exposed through the `start` function before it can be called. This is accomplished using a special object called `publicFacet`, which is returned by the `start` function. In the greeting contract, the `start` function exposes the `greet` function by defining it as a method of the contract's `publicFacet`, as shown below:

Reviewer 2: > must be exposed through the `start` function before it can be called.

The paragraph implies that the start function directly exposes other functions through publicFacet. However, the start function is primarily used to initialize the contract and return facets (such as publicFacet and creatorFacet), which can then expose methods like greet.

Author: Would it be fine if we removed "through the `start` function " from the first sentence? Note that the audience probably does not know anything about `facet` at this point.

Author: After our discussion, I am changing it to the following:

The `greet` function, along with any other public function, must be made accessible through the `publicFacet` of the contract. The `publicFacet` is returned by the `start` function, which initializes the contract. In the greeting contract, the `start` function exposes the `greet` function by defining it as a method of the contract's `publicFacet`, as shown below:

Reviewer: TECHNICAL:

"The publicFacet is returned by the start function" is incorrect. The start fuction returns an object with a `publicFacet` property.

Author: Fixed!

<<< @/../snippets/zoe/src/01-hello.js#publicFacet

> We wrap the reference to `greet` function in a `Far(...)` function call. This will allow callers to use it from outside the contract.

Reviewer 2: > We wrap the reference to `greet` function in a `Far(...)` function call. This will allow callers to use it from outside the contract.

I think it's a more accurate description of Far.

We wrap the greet function in a Far(...) call to safely expose it as a remote object, accessible from outside the contract.

Author: Will fix it.

We also give it a suggestive interface name `Hello` for debugging.
_We'll discuss [Far in more detail](../js-programming/far) later._

Putting it all together:

<<< @/../snippets/zoe/src/01-hello.js#contract

Let us save this code to a file named `01-hello.js` inside `src` directory. 
## Using, testing a contract

> Before we deploy or start our contract, it is usually a good idea to run a few tests to validate its correctness. Agoric contracts are typically tested using the [ava](https://github.com/avajs/ava) framework. The test file begins with an `import @endo/init` to establish a [Hardened JavaScript](../js-programming/hardened-js) environment as below:

Reviewer: TECHNICAL:

Why introduce deployment? Deploying this contract makes little sense. It doesn't have any offer handlers nor publish anything to vstorage, so there's no way to interact with it.

The PR title and description don't mention deployment, so it seems out of scope of this PR.

Author: @kbennett2000 the focus here is on explaining the deployment process, using a basic contract primarily as a means to that end, rather than the other way around.

> Before we deploy or start our contract, it is usually a good idea to run a few tests that validate its correctness. Agoric contracts are typically tested using the [ava](https://github.com/avajs/ava) framework. They start with `@endo/init` to establish a [Hardened JavaScript](../js-programming/hardened-js) environment as below:

Reviewer 2: Replace `They start with @endo/init...` with `Tests start with @endo/init...`

Replace `...few tests that validate...` with `...few tests to validate...`

Author: Fixed second one as suggested. Changed first one to `The test file begins with`.

<<< @/../snippets/zoe/contracts/test-zoe-01-hello.js#test-imports

> We also import `E()` in order to make asynchronous method calls and `test` function from `ava`. _We'll talk more about [using `E()` for async method calls](../js-programming/eventual-send) later._ Also note the `eslint-disable-next-line` comment on line 3; this is needed in order to suppress an error that arises due to a known issue in `ava`.

Reviewer: TECHNICAL:

Why should the reader note the eslint stuff? It seems unrelated to smart contract basics.

If they're curious, they can follow the URL.

I'd rather try to get rid of it than try to explain it here.

Reviewer: We could skip right to "A test that the `greet` method works..." and leave discussion of test set-up to the next section (complete walkthru).

Author: Fixed.

> We also import `E()` in order to make asynchronous method calls and `test` function from `ava`. _We'll talk more about [using `E()` for async method calls](../js-programming/eventual-send) later._ Also note the `eslint-disable-next-line` command in the comment on line 3; this is needed in order to suppress an error that arises due to a known issue in `ava`.

Reviewer 2: command -> comment

Author: Are you referring to the `eslint-disable-next-line`?  Let us talk about it. I think it should as is but I will fix it if you think otherwise.

Following these `import` statements, we write a simple test that validates that the `greet` method works as expected as below:

<<< @/../snippets/zoe/contracts/test-zoe-01-hello.js#test1

> Putting it all together:

Reviewer 2: I think "Putting it all together:" can be done in one go instead of showing the test code two times in same screen.

You can remove the test code (without import) above, and just have the following snippet.

Author: Fixed.

<<< @/../snippets/zoe/contracts/test-zoe-01-hello.js#test-01-hello

Let's save this code in a file named `test-01-hello.js` in a `test` directory. Both `src` and `test` directories should lie in the same `contract` directory. Let us run the following command to execute the test:

```sh

> npx ava --match="contract greets by name"

Reviewer: TECHNICAL (and **critical** due to security):

Let's not encourage use of `npx`. One typo like `npx aava` and who knows what code is running on their machine.

Author: Fixed.

> npx ava --match="contract greets by name"

Reviewer: The commands to run tests are discussed in the next section
https://docs.agoric.com/guides/zoe/contract-walkthru.html

Does it seem important to repeat them here?

Author: I do not see any significant overlap. Although, contract-walkthru needs a lot of fixing IMO but we will get to it once we have sorted out this and #1104.

```
You should see the following line towards the end of the output:
```
1 test passed
```

> Congratulations! You have written and tested your first smart contract. Our next goal is to deploy this contract on the local blockchain and make it more interesting by adding some business logic.

Reviewer: TECHNICAL:

Again, any discussion of deployment seems out of scope.

Reviewer: EDITORIAL:

Plus, the next goals are to learn about State and Access Control, no?

> Congratulations! You have written and tested your first smart contract. Our next goal is to deploy this contract on the local blockchain and make it more interesting by adding some business logic to it.

Reviewer 2: You can remove "to it" from the end.

Author: Fixed.

Let us save this code to a file named `01-hello.js` inside `src` directory. 
## Using, testing a contract

> Agoric contracts are typically tested using the [ava](https://github.com/avajs/ava) framework -  follow the instructions there to install `ava`. The test file begins with an `import @endo/init` to establish a [Hardened JavaScript](../js-programming/hardened-js) environment. We also import `E()` in order to make asynchronous method calls and `test` function from `ava`. _We'll talk more about [using `E()` for async method calls](../js-programming/eventual-send) later._ Following these `import` statements, we write a simple test that validates that the `greet` method works as expected.

Reviewer: Why "follow the instructions there to install `ava`"? In our getting started flow, `ava` is installed with other dev tools in the [yarn install step](https://docs.agoric.com/guides/getting-started/#installing-dapp-dependencies).

Author: I thought we may not assume that they have a `packages.json` file in their working directory?

Author: Should I change it to `yarn ava` ?

Reviewer: No `package.json`? Then how are dependencies such as `import '@endo/init';` satisfied? ES6 module `import` declarations won't even work without `"type": "module"` in `package.json`, right?

This is why I postponed details on how to run a test to the context of dapp-offer-up.

I didn't intend "contract basics" to be a tutorial where you could run the pieces. It was just a bit of background info between the getting-started tutorial and the contract walkthru.

Reviewer: Yes, if we're going to give a command, `yarn ava` is a good one to give.

Author: @dckc Pardon my ignorance in the above comment - they will of course have a `package.json`. I have updated the said command to `yarn ava` and removed the sentence: _"follow the instructions there to install `ava`"_. Thanks for your extensive and timely input.
